tools/fonts.R

```diff
@@ -4,41 +4,53 @@ if (Sys.getenv("RSTUDIO") == "1") {
+  matches <- regmatches(x, regexec(pattern, x))
+  na.omit(sapply(matches, "[", 2))
+}
+
 download_and_copy_fonts <-  function(theme) {
   theme_scss <- readLines(file.path(theme, "_bootswatch.scss"))
-  web_font_url <- grep("^\\$web-font-path: .*", theme_scss, value = TRUE)
-  if (length(web_font_url)) {
-    families <- sub("family=", "", str_extract(web_font_url, 'family=[^".]*'))
-    # some themes (sketchy) require multiple families
-    families <- unlist(strsplit(families, "|", fixed = TRUE))
-    families <- gsub("+", " ", families, fixed = TRUE)
-    families <- strsplit(families, ":")
-    families <- unlist(lapply(families, function(x) {
-      if (length(x) == 1) return(x)
-      paste0(x[[1]], ":", strsplit(x[[2]], ",")[[1]])
-    }))
-    # Assumes this repo is in the same directory, the shell script is executable,
-    # and the script's system dependencies are met
-    # https://github.com/neverpanic/google-font-download
-    # TODO: consider using this instead https://www.npmjs.com/package/grunt-google-fonts
-    cmd <- sprintf(
-      "../google-font-download/google-font-download -f ttf '%s'",
-      paste(families, collapse = "' '")
-    )
-    system(cmd)
-
-    # The css file generated by this tool assumes the font file resides in the
-    # same directory as the HTML file that this will eventually reside in,
-    # but it will make more sense to mount fonts in a fonts/ directory
-    font_css <- readLines("font.css")
-    font_css <- str_replace(font_css, "url\\('(.*\\.ttf)'\\)", "url('fonts/\\1')")
-    writeLines(font_css, "font.css")
-
-    file.rename("font.css", file.path(theme, "font.css"))
-    ttf_files <- Sys.glob("*.ttf")
-    dir.create(file.path("inst", "fonts"), showWarnings = FALSE)
-    file.rename(ttf_files, file.path("inst", "fonts", ttf_files))
+  web_font_path <- grep("$web-font-path:", theme_scss, value = TRUE, fixed = TRUE)
+  if (!length(web_font_path)) {
+    return()
+  }
+  web_font_url <- strsplit(stringr::str_trim(web_font_path), "\\s+")[[1]][2]
+  if (!length(web_font_url)) {
+    warning("Wasn't able to extract a url for $web-font-path from theme: ", theme)
+    return()
+  }
+  web_font_url <- gsub('(^")|("$)', '', web_font_url)
+
+  font_css <- httr::content(httr::GET(web_font_url))
+  font_css <- strsplit(font_css, "\n", fixed = TRUE)[[1]]
+
+  families <- extract_first_group(font_css, "font-family:\\s*'(.+)';")
+  if (length(families) == 0) {
+    stop("Expected to find at least one font-family")
   }
+  styles <- extract_first_group(font_css, "font-style:\\s*(.+);")
+  weights <- extract_first_group(font_css, "font-weight:\\s*(.+);")
+  font_ids <- paste0(
+    gsub("\\s+", "_", families), "_", weights,
+    ifelse(styles %in% "normal", "", styles)
+  )
+
+  # TODO: check urls length
+  local_files <- file.path(fonts_home, paste0(font_ids, ".ttf"))
+  urls <- unlist(stringr::str_extract_all(font_css, "https://fonts.gstatic.com/.*\\.ttf"))
+  download.file(urls, local_files)
```

Comment:
Extract whatever is in `url()` then use something like `tools::file_ext()` to construct the local filename